File: Android/app/src/main/java/com/nyaaydesk/app/presentation/auth/AuthViewModel.kt
package com.nyaaydesk.app.presentation.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

import com.nyaaydesk.app.data.repository.AuthRepository
import com.nyaaydesk.app.data.remote.dto.UserProfileDto
import io.github.jan.supabase.postgrest.postgrest

data class AuthUiState(
    val isLoading: Boolean = false,
    val isSuccess: Boolean = false,
    val errorMessage: String? = null,
    val userType: String? = null
)

/**
 * AuthViewModel — shared across all 4 login screens.
 *
 * The login function receives a `userType` hint but the actual role is extracted
 * from the JWT's user_metadata after login, ensuring the backend is the source of truth.
 */
@HiltViewModel
class AuthViewModel @Inject constructor(
    private val supabase: SupabaseClient,
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState

    fun login(email: String, password: String, expectedRole: String) {
        if (email.isBlank() || password.isBlank()) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please fill in all fields")
            return
        }

        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            try {
                supabase.auth.signInWith(Email) {
                    this.email = email
                    this.password = password
                }

                // 2. Fetch or Determine User Profile
                val session = supabase.auth.currentSessionOrNull()
                val metadata = session?.user?.userMetadata
                val userId = session?.user?.id
                var profile: UserProfileDto? = null

                if (userId != null) {
                    try {
                        profile = supabase.postgrest["users"]
                            .select { filter { eq("id", userId) } }
                            .decodeSingleOrNull<UserProfileDto>()
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }

                // Determine final userType
                val finalUserType = (profile?.userType ?: 
                                     (metadata?.get("user_type") ?: metadata?.get("role"))?.toString()?.trim('"') ?: 
                                     expectedRole).lowercase()

                // 3. Enforce role segregation
                if (finalUserType != expectedRole.lowercase()) {
                    supabase.auth.signOut()
                    _uiState.value = AuthUiState(
                        errorMessage = "Access denied. Your account is registered as ${finalUserType.replaceFirstChar { it.uppercase() }}, but you are trying to access the ${expectedRole} portal."
                    )
                    return@launch
                }

                // 4. Cache Session (This triggers dashboard ViewModels)
                if (profile != null) {
                    authRepository.saveUserSession(profile)
                } else if (userId != null) {
                    // Fallback profile if record is missing in 'users' table
                    val fallbackProfile = UserProfileDto(
                        id = userId,
                        email = email,
                        userType = expectedRole
                    )
                    authRepository.saveUserSession(fallbackProfile)
                }

                _uiState.value = AuthUiState(isSuccess = true, userType = expectedRole)

            } catch (e: Exception) {
                _uiState.value = AuthUiState(
                    errorMessage = when {
                        e.message?.contains("Invalid login") == true -> "Incorrect email or password."
                        e.message?.contains("Email not confirmed") == true -> "Please verify your email first."
                        else -> "Login failed: ${e.message ?: "Please check your internet connection"}"
                    }
                )
            }
        }
    }

    fun resetPassword(email: String) {
        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            try {
                supabase.auth.resetPasswordForEmail(email)
                _uiState.value = AuthUiState(errorMessage = "Password reset link sent to your email.")
            } catch (e: Exception) {
                _uiState.value = AuthUiState(errorMessage = "Failed to send reset link: ${e.message}")
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(errorMessage = null)
    }

    fun signOut() {
        viewModelScope.launch {
            runCatching { supabase.auth.signOut() }
        }
    }
}

File: IOS/Sources/NyaayDesk/Core/Auth/AuthViewModel.swift
import Foundation
import Observation
import Supabase

/**
 * AuthViewModel (iOS) — shared across all 4 login views.
 *
 * Uses Swift Concurrency (async/await) and the @Observable macro (iOS 17+)
 * to drive reactive SwiftUI updates.
 *
 * Role Enforcement: After login, extracts user_type from Supabase user_metadata.
 * If the role doesn't match the expected portal, the session is immediately invalidated.
 */
@Observable
final class AuthViewModel {
    var isLoading = false
    var isAuthenticated = false
    var errorMessage: String?
    var userType: String?
    var currentUser: UserProfile?

    private let supabase = SupabaseManager.shared.client

    // MARK: - Login
    func login(email: String, password: String, expectedRole: String) {
        guard !email.isEmpty, !password.isEmpty else {
            errorMessage = "Please fill in all fields"
            return
        }

        Task {
            isLoading = true
            errorMessage = nil
            defer { isLoading = false }

            do {
                let session = try await supabase.auth.signIn(email: email, password: password)

                // Extract user_type from Supabase user metadata
                let metadata = session.user.userMetadata
                var role = ""
                
                if let roleValue = metadata["user_type"], case .string(let r) = roleValue {
                    role = r
                } else if let roleValue = metadata["role"], case .string(let r) = roleValue {
                    role = r
                } else {
                    // Fallback to expected role if metadata is missing to avoid blocking devs/testers
                    role = expectedRole
                }

                // Enforce role segregation
                guard role.lowercased() == expectedRole.lowercased() else {
                    try await supabase.auth.signOut()
                    errorMessage = "Access denied. This portal is for \(expectedRole)s only."
                    return
                }

                let userId = session.user.id
                self.userType = role
                self.currentUser = UserProfile(
                    id: userId.uuidString,
                    email: email,
                    userType: role,
                    fullName: nil,
                    phone: nil,
                    barCouncilId: nil,
                    courtId: nil,
                    isVerified: false,
                    avatarUrl: nil
                )
                self.isAuthenticated = true

                // Background profile fetch - don't block the UI transition
                Task {
                    do {
                        let profile: UserProfile = try await supabase.database
                            .from("users")
                            .select()
                            .eq("id", value: userId)
                            .single()
                            .execute()
                            .value
                        await MainActor.run {
                            self.currentUser = profile
                        }
                    } catch {
                        print("Background profile fetch failed: \(error)")
                    }
                }

            } catch {
                if error.localizedDescription.contains("Invalid login") {
                    errorMessage = "Incorrect email or password."
                } else {
                    errorMessage = "Login failed: \(error.localizedDescription)"
                }
            }
        }
    }

    // MARK: - Password Reset
    func resetPassword(email: String) {
        Task {
            isLoading = true
            defer { isLoading = false }
            do {
                try await supabase.auth.resetPasswordForEmail(email)
                errorMessage = "Password reset link sent to your email."
            } catch {
                errorMessage = "Failed to send reset link: \(error.localizedDescription)"
            }
        }
    }

    // MARK: - Sign Out
    func signOut() {
        Task {
            try? await supabase.auth.signOut()
            isAuthenticated = false
            userType = nil
            currentUser = nil
            KeychainHelper.deleteSession()
        }
    }

    func clearError() {
        errorMessage = nil
    }
}
